### PR TITLE
Load dependencies and description from mod.conf

### PR DESCRIPTION
**Pull request by rubenwardy**
This deprecates depends.txt and description.txt

Inspired by #4539 by @TeTpaAka

### CODE REVIEW

File: builtin/mainmenu/modmgr.lua [hunk 1]
```diff
@@ -271,34 +271,79 @@ function modmgr.render_modlist(render_list)
...
+			mod_conf_contains_deps = true
+			local list = conf_soft_deps:split()
+			for i, element in ipairs(list) do
+				table.insert(soft_dependencies, element:trim())
+			end
+		end
+	end
+
+	-- Fallback to depends.txt
+	if not mod_conf_contains_deps then
```
Comment (rubenwardy) [inline builtin/mainmenu/modmgr.lua]: For backwards compatibility. Most mods currently have mod.conf with name

File: doc/lua_api.txt [hunk 2]
```diff
@@ -142,12 +141,25 @@ Mod directory structure
...
 
-
 ### modname
 The location of this directory can be fetched by using
 `minetest.get_modpath(modname)`.
 
+### mod.conf
+A key-value store of mod details.
+
+* `name` - the mod name. Allows Minetest to determine the mod name even if the folder is wrongly named.
```
Comment (rubenwardy) [inline doc/lua_api.txt]: It's an old feature that wasn't documented

File: builtin/mainmenu/modmgr.lua [hunk 3]
```diff
@@ -271,34 +271,79 @@ function modmgr.render_modlist(render_list)
...
+	local mod_conf = Settings(path .. DIR_DELIM .. "mod.conf")
+	local mod_conf_contains_deps = false
+	if mod_conf then
+		local conf_hard_deps = mod_conf:get("depends")
+		if conf_hard_deps then
+			mod_conf_contains_deps = true
+			local list = conf_hard_deps:split()
+			for i, element in ipairs(list) do
+				table.insert(hard_dependencies, element:trim())
+			end
```
Comment (rubenwardy) [inline builtin/mainmenu/modmgr.lua]: Good idea

File: src/mods.cpp [hunk 4]
```diff
@@ -60,28 +59,55 @@ void parseModContents(ModSpec &spec)
...
-			std::set<char> symbols;
-			if(parseDependsLine(is, dep, symbols)){
-				if(symbols.count('?') != 0){
-					spec.optdepends.insert(dep);
-				}
-				else{
-					spec.depends.insert(dep);
+
+	} else {
+		// Attempt to load dependencies from mod.conf
```
Comment (rubenwardy) [inline src/mods.cpp]: The alternative is to make a function and expose it to Lua. I'll have a look and see if if there's a good place for this. If there isn't, then for such small code it's not worth refactoring imo.

**Comment by rubenwardy** [comment]
I'd rather not bolt extra features in to this, in order to make it easier to merge. I don't mind striping   `[=><~]+[0-9.]+` to make backwards compatibility easier in the future. I also don't mind adding version to the docs, even if it's not used

File: builtin/mainmenu/modmgr.lua (hunk 1 above)
Comment (numberZero) [inline builtin/mainmenu/modmgr.lua]: Oh, okay then.

**Comment by numberZero** [comment]
Sounds reasonable.

File: doc/lua_api.txt [hunk 5]
```diff
@@ -142,12 +141,25 @@ Mod directory structure
...
+* `depends` - A comma separated list of dependencies. These are mods that must be loaded before this mod.
+* `optional_depends` - A comma separated list of optional dependencies. Like a dependency, but no error if the mod doesn't exist.
```
Comment (SmallJoker) [inline doc/lua_api.txt]: When we already get the opportunity - `dependencies` and `optional_dependencies` as nouns would fit better into the other settings we have.

File: src/mods.cpp [hunk 6]
```diff
@@ -60,28 +59,55 @@ void parseModContents(ModSpec &spec)
...
+		bool mod_conf_has_depends = false;
+		if (info.exists("depends")) {
+			mod_conf_has_depends = true;
+			std::string dep = info.get("depends");
+			for (const auto& dependency :  str_split(dep, ',')) {
```
Comment (SmallJoker) [inline src/mods.cpp]: `const auto &dependency : str_split(dep, ',')`, reference and superfluous space

Comment (rubenwardy) [inline src/mods.cpp]: oops, that was done automatically by IDE. (I fixed * in the code style, but not & it seems)

File: doc/lua_api.txt (hunk 5 above)
Comment (rubenwardy) [inline doc/lua_api.txt]: sounds good

**Comment by octacian** [comment]
:+1: However, with the prospect of the upcoming 0.5 release in view, I don't quite see the need for backward compatibility on this issue. Even as an extensive mod developer, I'd have no problem if there was no backward compatibility for this. For that matter, I'd like to see the `mod.conf` with a `name` attribute *required* for mods to load (and plan to do this in my modmgr rewrite). My point is that I don't think the extra code to maintain backward compatibility is worthwhile with 0.5 upcoming and the continued maintenance that would be required as a result.

**Comment by rubenwardy** [comment]
there's no point destroying compatibility for no reason

File: src/mods.cpp (hunk 6 above)
Comment (rubenwardy) [inline src/mods.cpp]: fixed

**Comment by SoniEx2** [comment]
Uh. Why not use JSON (with autodetection for backwards compatibility) and use a single deps array of objects?

```
"dependencies": [
  {"name": "thing"},
  {"name": "libthing", "version": "1.0.0,[1.1,1.6.1],1.6.2", "required": true}
]
```

**Comment by rubenwardy** [comment]
mod.conf already exists, and conf files are used elsewhere for package files (eg: game.conf)
I wouldn't be against supporting JSON, however I suggest making it very close to an existing format so they could be interchangable

**Comment by pgimeno** [comment]
> Even as an extensive mod developer, I'd have no problem if there was no backward compatibility for this.

My concern is unmaintained mods that stop working. For that reason, I'd like to keep backward compatibility.

File: src/mods.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rubenwardy) [reply src/mods.cpp]: fixed

**Comment by rubenwardy** [comment]
Fixed. The code is now a lot simpler and better, thanks for the suggestion @numberZero

File: src/mods.cpp [hunk 7]
```diff
@@ -60,28 +59,62 @@ void parseModContents(ModSpec &spec)
...
-					spec.depends.insert(dep);
+
+	} else {
+		// Attempt to load dependencies from mod.conf
+		bool mod_conf_has_depends = false;
+		if (info.exists("depends")) {
+			mod_conf_has_depends = true;
+			std::string dep = info.get("dependencies");
+			for (const auto &dependency : str_split(dep, ',')) {
+				spec.depends.insert(dependency);
```
Comment (SmallJoker) [inline src/mods.cpp]: Needs space trimming like the setting `secure.trusted_mods`: https://github.com/minetest/minetest/blob/master/src/script/lua_api/l_util.cpp#L407

File: src/mods.cpp [hunk 8]
```diff
@@ -60,28 +59,62 @@ void parseModContents(ModSpec &spec)
...
-				}
-				else{
-					spec.depends.insert(dep);
+
+	} else {
+		// Attempt to load dependencies from mod.conf
+		bool mod_conf_has_depends = false;
+		if (info.exists("depends")) {
+			mod_conf_has_depends = true;
+			std::string dep = info.get("dependencies");
```
Comment (SmallJoker) [inline src/mods.cpp]: Why not `get("depends")?

File: src/script/lua_api/l_mainmenu.cpp [hunk 9]
```diff
@@ -477,6 +427,103 @@ int ModApiMainMenu::l_delete_favorite(lua_State *L)
...
+int ModApiMainMenu::l_get_games(lua_State *L)
+{
+	std::vector<SubgameSpec> games = getAvailableGames();
+
+	lua_newtable(L);
+	int top = lua_gettop(L);
+	unsigned int index = 1;
+
+	for (const SubgameSpec &game : games) {
+		lua_pushnumber(L,index);
```
Comment (SmallJoker) [inline src/script/lua_api/l_mainmenu.cpp]: This touched function could do well with a few spaces after the commas and around the binary operators.

File: doc/lua_api.txt [hunk 10]
```diff
@@ -145,12 +144,29 @@ Mod directory structure
...
 The location of this directory can be fetched by using
 `minetest.get_modpath(modname)`.
 
+### mod.conf
+A key-value store of mod details.
+
+* `name` - the mod name. Allows Minetest to determine the mod name even if the folder is wrongly named.
+* `description` - Description of mod to be shown in the Mods tab of the mainmenu.
+* `depends` - A comma separated list of dependencies. These are mods that must be loaded before this mod.
+* `optional_depends` - A comma separated list of optional dependencies. Like a dependency, but no error if the mod doesn't exist.
```
Comment (SmallJoker) [inline doc/lua_api.txt]: These lines are a bit too long. Please split them.

**Comment by rubenwardy** [comment]
I've decided it's best to stick with `depends` in order to avoid typos

**Comment by rubenwardy** [comment]
just tested now. (Checked mods, configure, and loading the game with and w/o a required dep)

File: src/mods.cpp [hunk 11]
```diff
@@ -60,28 +60,66 @@ void parseModContents(ModSpec &spec)
...
+			std::vector<std::string> dependencies;
+
+			std::ifstream is((spec.path + DIR_DELIM + "depends.txt").c_str());
+			while (is.good()) {
+				std::string dep;
+				std::getline(is, dep);
+				dependencies.push_back(dep);
+			}
+
+			for (auto dependency : dependencies) {
```
Comment (nerzhul) [inline src/mods.cpp]: auto &

File: src/mods.cpp [hunk 12]
```diff
@@ -60,28 +60,66 @@ void parseModContents(ModSpec &spec)
...
+				std::set<char> symbols;
```
Comment (nerzhul) [inline src/mods.cpp]: is order important ? if not please use unordered set

**Comment by nerzhul** [comment]
Last, did you tested it in the CSM context ? I'm not sure but the most modified function can be used by CSM mod loader

File: doc/lua_api.txt [hunk 13]
```diff
@@ -145,12 +144,32 @@ Mod directory structure
...
 The location of this directory can be fetched by using
 `minetest.get_modpath(modname)`.
 
+### mod.conf
+A key-value store of mod details.
+
+* `name` - the mod name. Allows Minetest to determine the mod name even if the
+           folder is wrongly named.
+* `description` - Description of mod to be shown in the Mods tab of the mainmenu.
+* `depends` - A comma separated list of dependencies. These are mods that mus
```
Comment (rubenwardy) [inline doc/lua_api.txt]: t

**Comment by paramat** [comment]
> I've decided it's best to stick with depends in order to avoid typos

Yes good, 'dependencies' is long and a pain to type. 'depends' is familiar and helps with copypaste.

**Comment by rubenwardy** [comment]
I've not modified parseDependsString and it's only used by deprecated behaviour, so I don't see the value of a test. CSM will work fine.

Changed experimental to use mod.conf, loads fine

**Pull request merged**